server/src/main/java/org/opensearch/tiering/HotToWarmTieringService.java

```diff
@@ -0,0 +1,874 @@
+            clusterService.addListener(this);
+        }
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        // TODO add handling for master switch, dangling indices, master reload
+        if (event.routingTableChanged()) {
+            if (!indexTieringInfoMap.isEmpty()) {
+                updateIndexShardStatus(event.state());
+            }
+            if (!requestUuidToRequestContext.isEmpty()) {
+                completeRequestLevelTiering();
+            }
+        }
+    }
+
+    private void completeRequestLevelTiering() {
+        for (String requestUuid : requestUuidToRequestContext.keySet()) {
+            final HotToWarmTieringRequestContext requestContext = requestUuidToRequestContext.get(requestUuid);
+            if (requestContext.isRequestProcessingComplete()) {
+                logger.info("Tiering is completed for the request [{}]", requestContext);
+                for (Index index : indexTieringInfoMap.keySet()) {
+                    indexTieringInfoMap.remove(index);
+                }
+                requestUuidToRequestContext.remove(requestUuid);
+                if (requestContext.getRequest().waitForCompletion()) {
+                    requestContext.getListener().onResponse(requestContext.constructResponse());
+                }
+            }
+        }
+    }
+
+    private void processNotFoundIndex(IndexTieringInfo indexTieringInfo) {
+        Index index = indexTieringInfo.getIndex();
+        String reqId = indexTieringInfo.getRequestUuid();
+        HotToWarmTieringRequestContext requestContext = requestUuidToRequestContext.get(reqId);
+        requestContext.addToNotFound(index.getName());
+        requestContext.removeFromInProgress(index);
+        requestUuidToRequestContext.put(reqId, requestContext);
+        indexTieringInfoMap.remove(index);
+    }
+
+    protected boolean isSearchNode(DiscoveryNode node) {
+        return node != null && node.isSearchNode();
+    }
+
+    protected ShardTieringStatus updateShardTieringStatus(final ShardRouting shard, ClusterState clusterState) {
+        final boolean isShardFoundOnSearchNode = isSearchNode(clusterState.getNodes().get(shard.currentNodeId()));
+        final boolean isShardRelocatingToSearchNode = isSearchNode(clusterState.getNodes().get(shard.relocatingNodeId()));
+        ShardTieringState tieringState;
+        String reason = null;
+        if (shard.started() && isShardFoundOnSearchNode) {
+            tieringState = ShardTieringState.SUCCESSFUL;
+        } else if (shard.unassigned()) {
+            tieringState = ShardTieringState.INIT;
+            reason = "Shard is unassigned due to " + shard.unassignedInfo().getReason();
+        } else if (((shard.initializing() || shard.started()) && !isShardFoundOnSearchNode)
+            || (shard.relocating() && !isShardRelocatingToSearchNode)) {
```

Comment:
Is the `(shard.relocating() && !isShardRelocatingToSearchNode)` condition possible if there was an ongoing shard rebalance before the tiering request was submitted?

Reply:
that would not land up here as it would not be in the `indexTieiringInfoMap`